File: scripts/dzcto-learning.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "date"
require "fileutils"
require "json"
require "optparse"
require "rbconfig"
require "set"

TARGET_NEW_RATE = 0.35
RECENT_WINDOW = 12
REVIEW_PRESSURE_LIMIT = 3
STALE_REVIEW_DAYS = 3
INTERVALS = [1, 3, 7, 14, 30, 60].freeze

RATINGS = {
  "needs_work" => {
    label: "Needs Work",
    aliases: ["needs work", "needs_work", "work", "not", "no", "missed", "lost", "rough", "0"]
  },
  "familiar" => {
    label: "Familiar",
    aliases: ["familiar", "neutral", "partial", "some", "fuzzy", "1"]
  },
  "confident" => {
    label: "Confident",
    aliases: ["confident", "know", "solid", "clear", "got it", "2"]
  }
}.freeze

options = {
  date: Date.today.iso8601,
  mode: :select
}

OptionParser.new do |parser|
  parser.banner = "Usage: dzcto-learning.rb --project PATH [--select | --add | --record RATING]"

  parser.on("--project PATH", "Project folder; uses PATH/knowledge/wiki/learning") { |value| options[:project] = value }
  parser.on("--date DATE", "Date, YYYY-MM-DD") { |value| options[:date] = value }
  parser.on("--select", "Select a learning item or recommend adding a new one") { options[:mode] = :select }
  parser.on("--add", "Add a new learning item and mark it current") { options[:mode] = :add }
  parser.on("--record RATING", "Record rating: Needs Work, Familiar, or Confident") do |value|
    options[:mode] = :record
    options[:rating] = value
  end
  parser.on("--stats", "Print learning stats") { options[:mode] = :stats }
  parser.on("--id ID", "Learning item id") { |value| options[:id] = value }
  parser.on("--title TITLE", "Learning item title") { |value| options[:title] = value }
  parser.on("--summary SUMMARY", "One-sentence learning item summary") { |value| options[:summary] = value }
  parser.on("--details DETAILS", "Learning item explanation") { |value| options[:details] = value }
  parser.on("--details-file PATH", "File containing learning item explanation") { |value| options[:details_file] = value }
  parser.on("--source SOURCE", "Source file or artifact") { |value| options[:source] = value }
  parser.on("--tags TAGS", "Comma-separated tags") { |value| options[:tags] = value }
  parser.on("--note NOTE", "Optional review note") { |value| options[:note] = value }
end.parse!

abort "--project is required" unless options[:project]

def slugify(value)
  value.to_s.downcase.gsub(/[^a-z0-9]+/, "-").gsub(/^-|-$/, "")
end

def date_value(value)
  Date.iso8601(value.to_s)
rescue Date::Error
  nil
end

def wiki_root(project)
  File.join(File.expand_path(project), "knowledge", "wiki")
end

def learning_dir(project)
  File.join(wiki_root(project), "learning")
end

def items_path(project)
  File.join(learning_dir(project), "items.json")
end

def reviews_path(project)
  File.join(learning_dir(project), "reviews.jsonl")
end

def current_path(project)
  File.join(learning_dir(project), "current.json")
end

def ensure_learning_dir(project)
  FileUtils.mkdir_p(learning_dir(project))
end

def load_items(project)
  ensure_learning_dir(project)
  return [] unless File.exist?(items_path(project))

  items = JSON.parse(File.read(items_path(project)))
  items.is_a?(Array) ? items : []
rescue JSON::ParserError
  []
end

def save_items(project, items)
  ensure_learning_dir(project)
  File.write(items_path(project), "#{JSON.pretty_generate(items)}\n")
end

def load_reviews(project)
  ensure_learning_dir(project)
  return [] unless File.exist?(reviews_path(project))

  File.readlines(reviews_path(project), chomp: true).map do |line|
    JSON.parse(line)
  rescue JSON::ParserError
    nil
  end.compact
end

def append_review(project, review)
  ensure_learning_dir(project)
  File.open(reviews_path(project), "a") { |file| file.puts(JSON.generate(review)) }
end

def load_current(project)
  return nil unless File.exist?(current_path(project))

  JSON.parse(File.read(current_path(project)))
rescue JSON::ParserError
  nil
end

def save_current(project, item, kind, today)
  File.write(
    current_path(project),
    "#{JSON.pretty_generate({ id: item["id"], kind: kind, selected_on: today.iso8601 })}\n"
  )
end

def clear_current(project)
  FileUtils.rm_f(current_path(project))
end

def refresh_index(project)
  script = File.join(__dir__, "dzcto-artifact.rb")
  ok = system(RbConfig.ruby, script, "--project", File.expand_path(project), "--init", out: File::NULL)
  warn "Warning: failed to refresh wiki index" unless ok
end

def output(payload)
  puts JSON.pretty_generate(payload)
end

def rating_options
  RATINGS.map do |key, config|
    {
      rating: key,
      label: config[:label]
    }
  end
end

def normalize_rating(value)
  normalized = value.to_s.strip.downcase
  RATINGS.each do |key, config|
    return key if config[:aliases].include?(normalized)
  end

  nil
end

def active_items(items)
  items.select { |item| item.fetch("status", "active") == "active" }
end

def sort_new_items(items)
  items.sort_by { |item| [item["created_on"].to_s, item["title"].to_s.downcase] }
end

def sort_due_items(items, today)
  items.sort_by do |item|
    due_on = date_value(item["due_on"]) || today
    [due_on, item.fetch("box", 0).to_i, item["last_seen_on"].to_s, item["title"].to_s.downcase]
  end
end

def choose_learning_item(items, reviews, today)
  active = active_items(items)
  new_items = active.select { |item| item.fetch("seen_count", 0).to_i.zero? }
  due_items = active.select do |item|
    due_on = date_value(item["due_on"])
    item.fetch("seen_count", 0).to_i.positive? && due_on && due_on <= today
  end

  if due_items.empty? && new_items.empty?
    return {
      kind: "new_needed",
      reason: "No due review items or unseen learning items exist. Create one new system concept from current project context."
    }
  end

  review_pressure = due_items.length >= REVIEW_PRESSURE_LIMIT ||
                    due_items.any? { |item| (date_value(item["due_on"]) || today) <= today - STALE_REVIEW_DAYS }

  kind =
    if due_items.any? && review_pressure
      "review"
    elsif due_items.any? && new_items.any?
      recent = reviews.last(RECENT_WINDOW)
      recent_new_rate =
        if recent.empty?
          0.0
        else
          recent.count { |review| review["kind"] == "new" }.fdiv(recent.length)
        end

      recent_new_rate < TARGET_NEW_RATE ? "new" : "review"
    elsif due_items.any?
      "review"
    else
      "new"
    end

  item =
    if kind == "review"
      sort_due_items(due_items, today).first
    else
      sort_new_items(new_items).first
    end

  {
    kind: kind,
    item: item,
    reason: kind == "review" ? "Selected from due review items." : "Selected from unseen learning items."
  }
end

def unique_id(base, items)
  candidate = slugify(base)
  candidate = "learning-item" if candidate.empty?
  taken = items.map { |item| item["id"] }.to_set
  return candidate unless taken.include?(candidate)

  counter = 2
  loop do
    next_candidate = "#{candidate}-#{counter}"
    return next_candidate unless taken.include?(next_candidate)

    counter += 1
  end
end

def item_payload(item)
  {
    id: item["id"],
    title: item["title"],
    summary: item["summary"],
    details: item["details"],
    source: item["source"],
    tags: item["tags"] || [],
    box: item.fetch("box", 0),
    seen_count: item.fetch("seen_count", 0),
    due_on: item["due_on"],
    last_rating: item["last_rating"]
  }
end

def learning_stats(items, today)
  active = active_items(items)
  due = active.count do |item|
    due_on = date_value(item["due_on"])
    item.fetch("seen_count", 0).to_i.positive? && due_on && due_on <= today
  end
  unseen = active.count { |item| item.fetch("seen_count", 0).to_i.zero? }

  {
    active: active.length,
    due: due,
    new: unseen
  }
end

project = File.expand_path(options[:project])
today = Date.iso8601(options[:date])
ensure_learning_dir(project)

case options[:mode]
when :select
  items = load_items(project)
  reviews = load_reviews(project)
  selection = choose_learning_item(items, reviews, today)

  if selection[:item]
    save_current(project, selection[:item], selection[:kind], today)
    output(
      status: "selected",
      kind: selection[:kind],
      reason: selection[:reason],
      item: item_payload(selection[:item]),
      rating_options: rating_options,
      algorithm: "Review debt first; otherwise target #{((1 - TARGET_NEW_RATE) * 100).round}% review and #{(TARGET_NEW_RATE * 100).round}% new over the last #{RECENT_WINDOW} logged sessions."
    )
  else
    output(
      status: "new_needed",
      kind: "new",
      reason: selection[:reason],
      rating_options: rating_options,
      algorithm: "Review debt first; otherwise target #{((1 - TARGET_NEW_RATE) * 100).round}% review and #{(TARGET_NEW_RATE * 100).round}% new over the last #{RECENT_WINDOW} logged sessions."
    )
  end
when :add
  abort "--title is required with --add" unless options[:title]
  abort "--summary is required with --add" unless options[:summary]

  items = load_items(project)
  details =
    if options[:details_file]
      File.read(File.expand_path(options[:details_file]))
    else
      options[:details].to_s
    end
  id = options[:id] || unique_id(options[:title], items)
  tags = options[:tags].to_s.split(",").map(&:strip).reject(&:empty?)
  existing = items.find { |item| item["id"] == id }

  item = existing || {
    "id" => id,
    "created_on" => today.iso8601,
    "seen_count" => 0,
    "box" => 0,
    "status" => "active"
  }

  item.merge!(
    "title" => options[:title],
    "summary" => options[:summary],
    "details" => details,
    "source" => options[:source] || "Unknown",
    "tags" => tags,
    "due_on" => item["due_on"] || today.iso8601
  )

  items << item unless existing
  save_items(project, items)
  save_current(project, item, "new", today)
  refresh_index(project)

  output(
    status: existing ? "updated" : "added",
    kind: "new",
    item: item_payload(item),
    rating_options: rating_options
  )
when :record
  rating = normalize_rating(options[:rating])
  abort "Unknown rating '#{options[:rating]}'. Use Needs Work, Familiar, or Confident." unless rating

  items = load_items(project)
  current = load_current(project)
  id = options[:id] || current&.fetch("id", nil)
  abort "No current learning item. Run --select or pass --id." unless id

  item = items.find { |candidate| candidate["id"] == id }
  abort "Unknown learning item id '#{id}'" unless item

  previous_box = item.fetch("box", 0).to_i
  previous_seen_count = item.fetch("seen_count", 0).to_i
  kind = current&.fetch("kind", nil) || (previous_seen_count.zero? ? "new" : "review")

  next_box =
    case rating
    when "needs_work"
      [previous_box - 1, 0].max
    when "familiar"
      [previous_box + 1, INTERVALS.length - 1].min
    when "confident"
      [previous_box + 2, INTERVALS.length - 1].min
    end

  interval = rating == "needs_work" ? 1 : INTERVALS.fetch(next_box)
  due_on = today + interval
  item["box"] = next_box
  item["seen_count"] = previous_seen_count + 1
  item["last_seen_on"] = today.iso8601
  item["last_rating"] = RATINGS.fetch(rating)[:label]
  item["due_on"] = due_on.iso8601

  review = {
    reviewed_on: today.iso8601,
    id: item["id"],
    title: item["title"],
    kind: kind,
    rating: rating,
    rating_label: RATINGS.fetch(rating)[:label],
    previous_box: previous_box,
    next_box: next_box,
    previous_seen_count: previous_seen_count,
    next_seen_count: item["seen_count"],
    due_on: due_on.iso8601,
    note: options[:note]
  }.compact

  save_items(project, items)
  append_review(project, review)
  clear_current(project)
  refresh_index(project)

  output(
    status: "recorded",
    review: review,
    item: item_payload(item),
    stats: learning_stats(items, today)
  )
when :stats
  items = load_items(project)
  output(status: "stats", stats: learning_stats(items, today), rating_options: rating_options)
end

File: scripts/dzcto_learning.py
#!/usr/bin/env python3
"""Manage Day Zero CTO spaced-repetition learning state."""

from __future__ import annotations

import argparse
import datetime as dt
import json
import re
import subprocess
import sys
from pathlib import Path
from typing import Any


TARGET_NEW_RATE = 0.35
RECENT_WINDOW = 12
REVIEW_PRESSURE_LIMIT = 3
STALE_REVIEW_DAYS = 3
INTERVALS = [1, 3, 7, 14, 30, 60]

RATINGS = {
    "needs_work": {
        "label": "Needs Work",
        "aliases": {"needs work", "needs_work", "work", "not", "no", "missed", "lost", "rough", "0"},
    },
    "familiar": {
        "label": "Familiar",
        "aliases": {"familiar", "neutral", "partial", "some", "fuzzy", "1"},
    },
    "confident": {
        "label": "Confident",
        "aliases": {"confident", "know", "solid", "clear", "got it", "2"},
    },
}


def slugify(value: str) -> str:
    slug = re.sub(r"[^a-z0-9]+", "-", value.lower()).strip("-")
    return slug or "learning-item"


def date_value(value: Any) -> dt.date | None:
    try:
        return dt.date.fromisoformat(str(value))
    except (TypeError, ValueError):
        return None


def wiki_root(project: Path) -> Path:
    return project / "knowledge" / "wiki"


def learning_dir(project: Path) -> Path:
    return wiki_root(project) / "learning"


def items_path(project: Path) -> Path:
    return learning_dir(project) / "items.json"


def reviews_path(project: Path) -> Path:
    return learning_dir(project) / "reviews.jsonl"


def current_path(project: Path) -> Path:
    return learning_dir(project) / "current.json"


def checklist_dir(project: Path) -> Path:
    return learning_dir(project) / "checklists"


def checklist_path(project: Path) -> Path:
    return checklist_dir(project) / "mastery.md"


def ensure_learning_dir(project: Path) -> None:
    learning_dir(project).mkdir(parents=True, exist_ok=True)
    checklist_dir(project).mkdir(parents=True, exist_ok=True)


def read_json(path: Path, default: Any) -> Any:
    if not path.exists():
        return default
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except json.JSONDecodeError:
        return default


def load_items(project: Path) -> list[dict[str, Any]]:
    ensure_learning_dir(project)
    value = read_json(items_path(project), [])
    return value if isinstance(value, list) else []


def save_items(project: Path, items: list[dict[str, Any]]) -> None:
    ensure_learning_dir(project)
    items_path(project).write_text(json.dumps(items, indent=2) + "\n", encoding="utf-8")


def load_reviews(project: Path) -> list[dict[str, Any]]:
    ensure_learning_dir(project)
    path = reviews_path(project)
    if not path.exists():
        return []

    reviews: list[dict[str, Any]] = []
    for line in path.read_text(encoding="utf-8").splitlines():
        try:
            value = json.loads(line)
        except json.JSONDecodeError:
            continue
        if isinstance(value, dict):
            reviews.append(value)
    return reviews


def append_review(project: Path, review: dict[str, Any]) -> None:
    ensure_learning_dir(project)
    with reviews_path(project).open("a", encoding="utf-8") as file:
        file.write(json.dumps(review, separators=(",", ":")) + "\n")


def load_current(project: Path) -> dict[str, Any] | None:
    value = read_json(current_path(project), None)
    return value if isinstance(value, dict) else None


def save_current(project: Path, item: dict[str, Any], kind: str, today: dt.date) -> None:
    current_path(project).write_text(
        json.dumps({"id": item["id"], "kind": kind, "selected_on": today.isoformat()}, indent=2) + "\n",
        encoding="utf-8",
    )


def clear_current(project: Path) -> None:
    current_path(project).unlink(missing_ok=True)


def refresh_index(project: Path) -> None:
    script = Path(__file__).resolve().with_name("dzcto_artifact.py")
    result = subprocess.run(
        [sys.executable, str(script), "--project", str(project), "--init"],
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
        check=False,
    )
    if result.returncode != 0:
        print("Warning: failed to refresh wiki index", file=sys.stderr)


def rating_options() -> list[dict[str, str]]:
    return [{"rating": key, "label": config["label"]} for key, config in RATINGS.items()]


def normalize_rating(value: str | None) -> str | None:
    normalized = (value or "").strip().lower()
    for key, config in RATINGS.items():
        if normalized in config["aliases"]:
            return key
    return None


def active_items(items: list[dict[str, Any]]) -> list[dict[str, Any]]:
    return [item for item in items if item.get("status", "active") == "active"]


def sort_new_items(items: list[dict[str, Any]]) -> list[dict[str, Any]]:
    return sorted(items, key=lambda item: (str(item.get("created_on", "")), str(item.get("title", "")).lower()))


def sort_due_items(items: list[dict[str, Any]], today: dt.date) -> list[dict[str, Any]]:
    return sorted(
        items,
        key=lambda item: (
            date_value(item.get("due_on")) or today,
            int(item.get("box", 0) or 0),
            str(item.get("last_seen_on", "")),
            str(item.get("title", "")).lower(),
        ),
    )


def choose_learning_item(items: list[dict[str, Any]], reviews: list[dict[str, Any]], today: dt.date) -> dict[str, Any]:
    active = active_items(items)
    new_items = [item for item in active if int(item.get("seen_count", 0) or 0) == 0]
    due_items = []
    for item in active:
        due_on = date_value(item.get("due_on"))
        if int(item.get("seen_count", 0) or 0) > 0 and due_on and due_on <= today:
            due_items.append(item)

    if not due_items and not new_items:
        return {
            "kind": "new_needed",
            "reason": "No due review items or unseen learning items exist. Create one new system concept from current project context.",
        }

    review_pressure = len(due_items) >= REVIEW_PRESSURE_LIMIT or any(
        (date_value(item.get("due_on")) or today) <= today - dt.timedelta(days=STALE_REVIEW_DAYS)
        for item in due_items
    )

    if due_items and review_pressure:
        kind = "review"
    elif due_items and new_items:
        recent = reviews[-RECENT_WINDOW:]
        recent_new_rate = 0.0 if not recent else sum(1 for review in recent if review.get("kind") == "new") / len(recent)
        kind = "new" if recent_new_rate < TARGET_NEW_RATE else "review"
    elif due_items:
        kind = "review"
    else:
        kind = "new"

    item = sort_due_items(due_items, today)[0] if kind == "review" else sort_new_items(new_items)[0]
    return {
        "kind": kind,
        "item": item,
        "reason": "Selected from due review items." if kind == "review" else "Selected from unseen learning items.",
    }


def unique_id(base: str, items: list[dict[str, Any]]) -> str:
    candidate = slugify(base)
    taken = {item.get("id") for item in items}
    if candidate not in taken:
        return candidate

    counter = 2
    while True:
        next_candidate = f"{candidate}-{counter}"
        if next_candidate not in taken:
            return next_candidate
        counter += 1


def item_payload(item: dict[str, Any]) -> dict[str, Any]:
    return {
        "id": item.get("id"),
        "title": item.get("title"),
        "summary": item.get("summary"),
        "details": item.get("details"),
        "source": item.get("source"),
        "tags": item.get("tags") or [],
        "box": item.get("box", 0),
        "seen_count": item.get("seen_count", 0),
        "due_on": item.get("due_on"),
        "last_rating": item.get("last_rating"),
        "mastered_on": item.get("mastered_on"),
    }


def learning_stats(items: list[dict[str, Any]], today: dt.date) -> dict[str, int]:
    active = active_items(items)
    due = 0
    for item in active:
        due_on = date_value(item.get("due_on"))
        if int(item.get("seen_count", 0) or 0) > 0 and due_on and due_on <= today:
            due += 1
    unseen = sum(1 for item in active if int(item.get("seen_count", 0) or 0) == 0)
    mastered = sum(1 for item in active if item.get("mastered_on"))
    return {"active": len(active), "due": due, "new": unseen, "mastered": mastered}


def write_mastery_checklist(project: Path, items: list[dict[str, Any]], today: dt.date) -> dict[str, Any]:
    ensure_learning_dir(project)
    active = sorted(active_items(items), key=lambda item: str(item.get("title", "")).lower())
    total = len(active)
    confirmed = sum(1 for item in active if item.get("mastered_on"))

    lines = [
        "---",
        "source: Day Zero CTO learning",
        f"updated_on: {today.isoformat()}",
        "---",
        "",
        "# Learning Mastery Checklist",
        "",
        f"Progress: {confirmed}/{total} concepts confirmed",
        "",
        "## Active Concepts",
        "",
    ]

    if active:
        for item in active:
            mark = "x" if item.get("mastered_on") else " "
            source = item.get("source") or "Unknown"
            lines.append(f"- [{mark}] {item.get('title')} (`{item.get('id')}`) - {source}")
    else:
        lines.append("_No active learning items yet._")

    lines.append("")
    checklist_path(project).write_text("\n".join(lines), encoding="utf-8")
    percent = round((confirmed / total) * 100) if total else 0
    return {
        "path": str(checklist_path(project)),
        "confirmed": confirmed,
        "total": total,
        "percent": percent,
    }


def add_or_update_item(
    items: list[dict[str, Any]],
    today: dt.date,
    *,
    item_id: str | None,
    title: str,
    summary: str,
    details: str,
    source: str | None,
    tags: list[str],
    due_on: str | None = None,
    created_on: str | None = None,
) -> tuple[dict[str, Any], bool]:
    actual_id = item_id or unique_id(title, items)
    existing = next((item for item in items if item.get("id") == actual_id), None)
    item = existing or {
        "id": actual_id,
        "created_on": created_on or today.isoformat(),
        "seen_count": 0,
        "box": 0,
        "status": "active",
    }

    item.update(
        {
            "title": title,
            "summary": summary,
            "details": details or summary,
            "source": source or "Unknown",
            "tags": tags,
            "due_on": item.get("due_on") or due_on or today.isoformat(),
        }
    )

    if not existing:
        items.append(item)
    return item, existing is not None


def parse_tags(value: Any) -> list[str]:
    if isinstance(value, list):
        return [str(tag).strip() for tag in value if str(tag).strip()]
    return [tag.strip() for tag in str(value or "").split(",") if tag.strip()]


def compact(value: dict[str, Any]) -> dict[str, Any]:
    return {key: item for key, item in value.items() if item is not None}


def output(payload: dict[str, Any]) -> None:
    print(json.dumps(payload, indent=2))


def main(argv: list[str]) -> int:
    parser = argparse.ArgumentParser(description="Manage Day Zero CTO learning")
    parser.add_argument("--project", required=True, help="Project folder; uses PATH/knowledge/wiki/learning")
    parser.add_argument("--date", default=dt.date.today().isoformat())
    mode = parser.add_mutually_exclusive_group()
    mode.add_argument("--select", action="store_true", help="Select a learning item")
    mode.add_argument("--add", action="store_true", help="Add a learning item")
    mode.add_argument("--seed-file", help="Add multiple learning items from a JSON array")
    mode.add_argument("--record", help="Record rating: Needs Work, Familiar, or Confident")
    mode.add_argument("--stats", action="store_true", help="Print learning stats")
    parser.add_argument("--id", help="Learning item id")
    parser.add_argument("--title", help="Learning item title")
    parser.add_argument("--summary", help="One-sentence learning item summary")
    parser.add_argument("--details", help="Learning item explanation")
    parser.add_argument("--details-file", help="File containing learning item explanation")
    parser.add_argument("--source", help="Source file or artifact")
    parser.add_argument("--tags", help="Comma-separated tags")
    parser.add_argument("--note", help="Optional review note")
    args = parser.parse_args(argv)

    project = Path(args.project).expanduser().resolve()
    # Operating on a project whose wiki does not exist is almost always an agent
    # passing the company NAME instead of the project FOLDER path; fail loudly
    # rather than silently creating a stray directory. (Onboarding runs `dzcto
    # init` first, so the wiki already exists by the time learning items are seeded.)
    if not wiki_root(project).exists():
        raise SystemExit(
            f"No Day Zero CTO wiki found at {wiki_root(project)}.\n"
            f"--project must be the project FOLDER path (for example ~/Documents/Acme CTO), "
            f"not a company name.\n"
            f"If this project has not been set up yet, run `dzcto init` on the project folder first."
        )
    today = dt.date.fromisoformat(args.date)
    ensure_learning_dir(project)

    if args.add:
        if not args.title:
            raise SystemExit("--title is required with --add")
        if not args.summary:
            raise SystemExit("--summary is required with --add")

        items = load_items(project)
        details = Path(args.details_file).expanduser().read_text(encoding="utf-8") if args.details_file else (args.details or "")
        item, existed = add_or_update_item(
            items,
            today,
            item_id=args.id,
            title=args.title,
            summary=args.summary,
            details=details,
            source=args.source,
            tags=parse_tags(args.tags),
        )
        save_items(project, items)
        save_current(project, item, "new", today)
        checklist = write_mastery_checklist(project, items, today)
        refresh_index(project)
        output(
            {
                "status": "updated" if existed else "added",
                "kind": "new",
                "item": item_payload(item),
                "checklist": checklist,
                "rating_options": rating_options(),
            }
        )
        return 0

    if args.seed_file:
        seed_items = json.loads(Path(args.seed_file).expanduser().read_text(encoding="utf-8"))
        if not isinstance(seed_items, list):
            raise SystemExit("--seed-file must contain a JSON array")

        items = load_items(project)
        added: list[dict[str, Any]] = []
        updated: list[dict[str, Any]] = []
        for seed in seed_items:
            if not isinstance(seed, dict):
                continue
            title = str(seed.get("title") or "").strip()
            summary = str(seed.get("summary") or "").strip()
            if not title or not summary:
                continue
            item, existed = add_or_update_item(
                items,
                today,
                item_id=str(seed.get("id") or "").strip() or None,
                title=title,
                summary=summary,
                details=str(seed.get("details") or summary),
                source=str(seed.get("source") or "Onboarding"),
                tags=parse_tags(seed.get("tags")),
                due_on=str(seed.get("due_on") or today.isoformat()),
                created_on=str(seed.get("created_on") or today.isoformat()),
            )
            updated.append(item) if existed else added.append(item)

        save_items(project, items)
        checklist = write_mastery_checklist(project, items, today)
        refresh_index(project)
        output(
            {
                "status": "seeded",
                "added": len(added),
                "updated": len(updated),
                "items": [item_payload(item) for item in added + updated],
                "stats": learning_stats(items, today),
                "checklist": checklist,
                "rating_options": rating_options(),
            }
        )
        return 0

    if args.record:
        rating = normalize_rating(args.record)
        if not rating:
            raise SystemExit(f"Unknown rating '{args.record}'. Use Needs Work, Familiar, or Confident.")

        items = load_items(project)
        current = load_current(project)
        item_id = args.id or (current or {}).get("id")
        if not item_id:
            raise SystemExit("No current learning item. Run --select or pass --id.")

        item = next((candidate for candidate in items if candidate.get("id") == item_id), None)
        if not item:
            raise SystemExit(f"Unknown learning item id '{item_id}'")

        previous_box = int(item.get("box", 0) or 0)
        previous_seen_count = int(item.get("seen_count", 0) or 0)
        kind = (current or {}).get("kind") or ("new" if previous_seen_count == 0 else "review")

        if rating == "needs_work":
            next_box = max(previous_box - 1, 0)
        elif rating == "familiar":
            next_box = min(previous_box + 1, len(INTERVALS) - 1)
        else:
            next_box = min(previous_box + 2, len(INTERVALS) - 1)

        interval = 1 if rating == "needs_work" else INTERVALS[next_box]
        due_on = today + dt.timedelta(days=interval)
        item["box"] = next_box
        item["seen_count"] = previous_seen_count + 1
        item["last_seen_on"] = today.isoformat()
        item["last_rating"] = RATINGS[rating]["label"]
        item["due_on"] = due_on.isoformat()
        if rating == "confident":
            item["mastered_on"] = item.get("mastered_on") or today.isoformat()
        else:
            item.pop("mastered_on", None)

        review = compact(
            {
                "reviewed_on": today.isoformat(),
                "id": item.get("id"),
                "title": item.get("title"),
                "kind": kind,
                "rating": rating,
                "rating_label": RATINGS[rating]["label"],
                "previous_box": previous_box,
                "next_box": next_box,
                "previous_seen_count": previous_seen_count,
                "next_seen_count": item["seen_count"],
                "due_on": due_on.isoformat(),
                "mastered": rating == "confident",
                "note": args.note,
            }
        )

        save_items(project, items)
        append_review(project, review)
        clear_current(project)
        checklist = write_mastery_checklist(project, items, today)
        refresh_index(project)
        output({"status": "recorded", "review": review, "item": item_payload(item), "stats": learning_stats(items, today), "checklist": checklist})
        return 0

    if args.stats:
        items = load_items(project)
        checklist = write_mastery_checklist(project, items, today)
        refresh_index(project)
        output({"status": "stats", "stats": learning_stats(items, today), "checklist": checklist, "rating_options": rating_options()})
        return 0

    items = load_items(project)
    reviews = load_reviews(project)
    checklist = write_mastery_checklist(project, items, today)
    selection = choose_learning_item(items, reviews, today)
    algorithm = f"Review debt first; otherwise target {round((1 - TARGET_NEW_RATE) * 100)}% review and {round(TARGET_NEW_RATE * 100)}% new over the last {RECENT_WINDOW} logged sessions."

    if selection.get("item"):
        save_current(project, selection["item"], selection["kind"], today)
        refresh_index(project)
        output(
            {
                "status": "selected",
                "kind": selection["kind"],
                "reason": selection["reason"],
                "item": item_payload(selection["item"]),
                "checklist": checklist,
                "rating_options": rating_options(),
                "algorithm": algorithm,
            }
        )
    else:
        refresh_index(project)
        output(
            {
                "status": "new_needed",
                "kind": "new",
                "reason": selection["reason"],
                "checklist": checklist,
                "rating_options": rating_options(),
                "algorithm": algorithm,
            }
        )
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))
